File: android/app/src/main/java/io/sekretess/cryptography/storage/SekretessSignalProtocolStore.java
package io.sekretess.cryptography.storage;

import android.util.Log;

import org.signal.libsignal.protocol.IdentityKey;
import org.signal.libsignal.protocol.IdentityKeyPair;
import org.signal.libsignal.protocol.InvalidKeyIdException;
import org.signal.libsignal.protocol.NoSessionException;
import org.signal.libsignal.protocol.SignalProtocolAddress;
import org.signal.libsignal.protocol.groups.state.SenderKeyRecord;
import org.signal.libsignal.protocol.state.KyberPreKeyRecord;
import org.signal.libsignal.protocol.state.PreKeyRecord;
import org.signal.libsignal.protocol.state.SessionRecord;
import org.signal.libsignal.protocol.state.SignalProtocolStore;
import org.signal.libsignal.protocol.state.SignedPreKeyRecord;

import java.util.List;
import java.util.UUID;

import io.sekretess.db.repository.IdentityKeyRepository;
import io.sekretess.db.repository.KyberPreKeyRepository;
import io.sekretess.db.repository.PreKeyRepository;
import io.sekretess.db.repository.RegistrationRepository;
import io.sekretess.db.repository.SenderKeyRepository;
import io.sekretess.db.repository.SessionRepository;
import io.sekretess.db.repository.SignedPreKeyRepository;

public class SekretessSignalProtocolStore implements SignalProtocolStore {
    private final SekretessIdentityKeyStore identityKeyStore;
    private final SekretessPreKeyStore preKeyStore;
    private final SekretessSessionStore sessionStore;

    private final SekretessSignedPreKeyStore signedPreKeyStore;

    private final SekretessSenderKeyStore senderKeyStore;

    private final SekretessKyberPreKeyStore kyberPreKeyStore;
    private int minKeysThreshold = 5;

    public SekretessSignalProtocolStore(IdentityKeyRepository identityKeyRepository,
                                        RegistrationRepository registrationRepository,
                                        PreKeyRepository preKeyRepository,
                                        SignedPreKeyRepository signedPreKeyRepository,
                                        SessionRepository sessionRepository,
                                        SenderKeyRepository senderKeyRepository,
                                        KyberPreKeyRepository kyberPreKeyRepository) {
        this.identityKeyStore = new SekretessIdentityKeyStore(identityKeyRepository, registrationRepository);
        this.preKeyStore = new SekretessPreKeyStore(preKeyRepository);
        this.sessionStore = new SekretessSessionStore(sessionRepository);
        this.signedPreKeyStore = new SekretessSignedPreKeyStore(preKeyRepository, signedPreKeyRepository);
        this.senderKeyStore = new SekretessSenderKeyStore(senderKeyRepository);
        this.kyberPreKeyStore = new SekretessKyberPreKeyStore(kyberPreKeyRepository);


        Log.i("SekretessSignalProtocolStore", "SignalProtocolStore initialized");
    }

    public boolean registrationRequired() {
        boolean registrationRequired = identityKeyStore.registrationRequired();
        Log.i("SekretessSignalProtocolStore", "registrationRequired: " + registrationRequired);
        return registrationRequired;
    }

    @Override
    public void storeSenderKey(SignalProtocolAddress sender, UUID distributionId, SenderKeyRecord record) {
        senderKeyStore.storeSenderKey(sender, distributionId, record);
    }

    @Override
    public SenderKeyRecord loadSenderKey(SignalProtocolAddress sender, UUID distributionId) {
        return senderKeyStore.loadSenderKey(sender, distributionId);
    }

    @Override
    public IdentityKeyPair getIdentityKeyPair() {
        return identityKeyStore.getIdentityKeyPair();
    }

    @Override
    public int getLocalRegistrationId() {
        return identityKeyStore.getLocalRegistrationId();
    }

    @Override
    public IdentityChange saveIdentity(SignalProtocolAddress address, IdentityKey identityKey) {
        return identityKeyStore.saveIdentity(address, identityKey);
    }

    @Override
    public boolean isTrustedIdentity(SignalProtocolAddress address, IdentityKey identityKey, Direction direction) {
        return identityKeyStore.isTrustedIdentity(address, identityKey, direction);
    }

    @Override
    public IdentityKey getIdentity(SignalProtocolAddress address) {
        return identityKeyStore.getIdentity(address);
    }

    @Override
    public KyberPreKeyRecord loadKyberPreKey(int kyberPreKeyId) throws InvalidKeyIdException {
        return kyberPreKeyStore.loadKyberPreKey(kyberPreKeyId);
    }

    @Override
    public List<KyberPreKeyRecord> loadKyberPreKeys() {
        return kyberPreKeyStore.loadKyberPreKeys();
    }

    @Override
    public void storeKyberPreKey(int kyberPreKeyId, KyberPreKeyRecord record) {
        kyberPreKeyStore.storeKyberPreKey(kyberPreKeyId, record);
    }

    @Override
    public boolean containsKyberPreKey(int kyberPreKeyId) {
        return kyberPreKeyStore.containsKyberPreKey(kyberPreKeyId);
    }

    @Override
    public void markKyberPreKeyUsed(int kyberPreKeyId) {
        kyberPreKeyStore.markKyberPreKeyUsed(kyberPreKeyId);
    }

    @Override
    public PreKeyRecord loadPreKey(int preKeyId) throws InvalidKeyIdException {
        return preKeyStore.loadPreKey(preKeyId);
    }

    @Override
    public void storePreKey(int preKeyId, PreKeyRecord record) {
        preKeyStore.storePreKey(preKeyId, record);
    }

    @Override
    public boolean containsPreKey(int preKeyId) {
        return preKeyStore.containsPreKey(preKeyId);
    }

    @Override
    public void removePreKey(int preKeyId) {
        preKeyStore.removePreKey(preKeyId);
    }

    @Override
    public SessionRecord loadSession(SignalProtocolAddress address) {
        return sessionStore.loadSession(address);
    }

    @Override
    public List<SessionRecord> loadExistingSessions(List<SignalProtocolAddress> addresses) throws NoSessionException {
        return sessionStore.loadExistingSessions(addresses);
    }

    @Override
    public List<Integer> getSubDeviceSessions(String name) {
        return sessionStore.getSubDeviceSessions(name);
    }

    @Override
    public void storeSession(SignalProtocolAddress address, SessionRecord record) {
        sessionStore.storeSession(address, record);
    }

    @Override
    public boolean containsSession(SignalProtocolAddress address) {
        return sessionStore.containsSession(address);
    }

    @Override
    public void deleteSession(SignalProtocolAddress address) {
        sessionStore.deleteSession(address);
    }

    @Override
    public void deleteAllSessions(String name) {
        sessionStore.deleteAllSessions(name);
    }

    @Override
    public SignedPreKeyRecord loadSignedPreKey(int signedPreKeyId) throws InvalidKeyIdException {
        return signedPreKeyStore.loadSignedPreKey(signedPreKeyId);
    }

    @Override
    public List<SignedPreKeyRecord> loadSignedPreKeys() {
        return signedPreKeyStore.loadSignedPreKeys();
    }

    @Override
    public void storeSignedPreKey(int signedPreKeyId, SignedPreKeyRecord record) {
        signedPreKeyStore.storeSignedPreKey(signedPreKeyId, record);
    }

    @Override
    public boolean containsSignedPreKey(int signedPreKeyId) {
        return signedPreKeyStore.containsSignedPreKey(signedPreKeyId);
    }

    @Override
    public void removeSignedPreKey(int signedPreKeyId) {
        signedPreKeyStore.removeSignedPreKey(signedPreKeyId);
    }

    public boolean clearStorage() {
        this.identityKeyStore.clearStorage();
        this.preKeyStore.clearStorage();
        this.sessionStore.clearStorage();
        this.signedPreKeyStore.clearStorage();
        this.senderKeyStore.clearStorage();
        this.kyberPreKeyStore.clearStorage();
        return true;
    }


    public boolean updateKeysRequired() {
        return preKeyStore.count() <= minKeysThreshold;
    }
}

File: ios/Runner/SekretessSignalProtocolStore.swift
import Foundation
import LibSignalClient

/// Signal Protocol Store implementation for iOS.
/// Aggregates all Signal Protocol stores and manages all Signal Protocol state.
class SekretessSignalProtocolStore {
    let identityKeyStore: SekretessIdentityKeyStore
    let preKeyStore: SekretessPreKeyStore
    let sessionStore: SekretessSessionStore
    let signedPreKeyStore: SekretessSignedPreKeyStore
    let senderKeyStore: SekretessSenderKeyStore
    let kyberPreKeyStore: SekretessKyberPreKeyStore
    private let minKeysThreshold = 5
    
    init(
        identityKeyRepository: IdentityKeyRepository,
        registrationRepository: RegistrationRepository,
        preKeyRepository: PreKeyRepository,
        signedPreKeyRepository: SignedPreKeyRepository,
        sessionRepository: SessionRepository,
        senderKeyRepository: SenderKeyRepository,
        kyberPreKeyRepository: KyberPreKeyRepository
    ) {
        self.identityKeyStore = SekretessIdentityKeyStore(
            identityKeyRepository: identityKeyRepository,
            registrationRepository: registrationRepository
        )
        self.preKeyStore = SekretessPreKeyStore(preKeyRepository: preKeyRepository)
        self.sessionStore = SekretessSessionStore(sessionRepository: sessionRepository)
        self.signedPreKeyStore = SekretessSignedPreKeyStore(
            preKeyRepository: preKeyRepository,
            signedPreKeyRepository: signedPreKeyRepository
        )
        self.senderKeyStore = SekretessSenderKeyStore(senderKeyRepository: senderKeyRepository)
        self.kyberPreKeyStore = SekretessKyberPreKeyStore(kyberPreKeyRepository: kyberPreKeyRepository)
        
        print("SekretessSignalProtocolStore: SignalProtocolStore initialized")
    }
    
    func registrationRequired() -> Bool {
        let required = identityKeyStore.registrationRequired()
        print("SekretessSignalProtocolStore: registrationRequired: \(required)")
        return required
    }
    
    func updateKeysRequired() -> Bool {
        return preKeyStore.count() <= minKeysThreshold
    }
    
    func clearStorage() {
        identityKeyStore.clearStorage()
        preKeyStore.clearStorage()
        sessionStore.clearStorage()
        signedPreKeyStore.clearStorage()
        senderKeyStore.clearStorage()
        kyberPreKeyStore.clearStorage()
    }
    
    // MARK: - SignalProtocolStore Protocol Implementation
    
    func getIdentityKeyPair() -> IdentityKeyPair {
        return identityKeyStore.getIdentityKeyPair()
    }
    
    func getLocalRegistrationId() -> UInt32 {
        return identityKeyStore.getLocalRegistrationId()
    }
    
    func saveIdentity(_ address: ProtocolAddress, _ identity: IdentityKey) -> IdentityChange {
        return identityKeyStore.saveIdentity(address, identity)
    }
    
    func isTrustedIdentity(_ address: ProtocolAddress, _ identity: IdentityKey, _ direction: Direction) -> Bool {
        return identityKeyStore.isTrustedIdentity(address, identity, direction)
    }
    
    func getIdentity(_ address: ProtocolAddress) -> IdentityKey? {
        return identityKeyStore.getIdentity(address)
    }
    
    func loadPreKey(id: UInt32) throws -> PreKeyRecord {
        return try preKeyStore.loadPreKey(id: id)
    }
    
    func storePreKey(_ id: UInt32, _ record: PreKeyRecord) {
        preKeyStore.storePreKey(id, record)
    }
    
    func removePreKey(_ id: UInt32) {
        preKeyStore.removePreKey(id)
    }
    
    func loadSignedPreKey(id: UInt32) throws -> SignedPreKeyRecord {
        return try signedPreKeyStore.loadSignedPreKey(id: id)
    }
    
    func storeSignedPreKey(_ id: UInt32, _ record: SignedPreKeyRecord) {
        signedPreKeyStore.storeSignedPreKey(id, record)
    }
    
    func removeSignedPreKey(_ id: UInt32) {
        signedPreKeyStore.removeSignedPreKey(id)
    }
    
    func loadKyberPreKey(id: UInt32) throws -> KyberPreKeyRecord {
        return try kyberPreKeyStore.loadKyberPreKey(id: id)
    }
    
    func storeKyberPreKey(_ id: UInt32, _ record: KyberPreKeyRecord) {
        kyberPreKeyStore.storeKyberPreKey(id, record)
    }
    
    func removeKyberPreKey(_ id: UInt32) {
        kyberPreKeyStore.removeKyberPreKey(id)
    }
    
    func loadKyberPreKeys() -> [KyberPreKeyRecord] {
        return kyberPreKeyStore.loadKyberPreKeys()
    }
    
    func containsKyberPreKey(_ id: UInt32) -> Bool {
        return kyberPreKeyStore.containsKyberPreKey(id)
    }
    
    func markKyberPreKeyUsed(_ id: UInt32) {
        kyberPreKeyStore.markKyberPreKeyUsed(id)
    }
    
    func loadSession(for address: ProtocolAddress) -> SessionRecord? {
        return try? sessionStore.loadSession(for: address, context: SimpleStoreContext())
    }
    
    func loadExistingSessions(for addresses: [ProtocolAddress]) throws -> [SessionRecord] {
        return try sessionStore.loadExistingSessions(for: addresses, context: SimpleStoreContext())
    }
    
    func getSubDeviceSessions(name: String) -> [UInt32] {
        return sessionStore.getSubDeviceSessions(name: name)
    }
    
    func containsSession(_ address: ProtocolAddress) -> Bool {
        return sessionStore.containsSession(address)
    }
    
    func deleteAllSessions(name: String) {
        sessionStore.deleteAllSessions(name: name)
    }
    
    func loadSignedPreKeys() -> [SignedPreKeyRecord] {
        return signedPreKeyStore.loadSignedPreKeys()
    }
    
    func containsSignedPreKey(_ id: UInt32) -> Bool {
        return signedPreKeyStore.containsSignedPreKey(id)
    }
    
    func containsPreKey(_ id: UInt32) -> Bool {
        return preKeyStore.containsPreKey(id)
    }
    
    func storeSession(_ address: ProtocolAddress, _ record: SessionRecord) {
        do {
            try sessionStore.storeSession(record, for: address, context: SimpleStoreContext())
        } catch {
            print("Error storing session: \(error)")
        }
    }
    
    func removeSession(for address: ProtocolAddress) {
        sessionStore.removeSession(for: address)
    }
    
    func loadSenderKey(for distributionId: UUID, senderName: String, deviceId: UInt32) -> SenderKeyRecord? {
        do {
            let address = try ProtocolAddress(name: senderName, deviceId: deviceId)
            return try senderKeyStore.loadSenderKey(from: address, distributionId: distributionId, context: SimpleStoreContext())
        } catch {
            print("Error loading sender key: \(error)")
            return nil
        }
    }
    
    func storeSenderKey(_ distributionId: UUID, _ address: ProtocolAddress, _ record: SenderKeyRecord) {
        do {
            try senderKeyStore.storeSenderKey(from: address, distributionId: distributionId, record: record, context: SimpleStoreContext())
        } catch {
            print("Error storing sender key: \(error)")
        }
    }
}
